R/db-odbc-teradata.R

```diff
@@ -0,0 +1,94 @@
+#' @export
+sql_select.Teradata <- function(con, select, from, where = NULL,
+                                             group_by = NULL, having = NULL,
+                                             order_by = NULL,
+                                             limit = NULL,
+                                             distinct = FALSE,
+                                             ...) {
+  out <- vector("list", 7)
+  names(out) <- c("select", "from", "where", "group_by",
+                  "having", "order_by","limit")
+
+  assert_that(is.character(select), length(select) > 0L)
+  out$select    <- build_sql(
+    "SELECT ",
+
+    if (distinct) sql("DISTINCT "),
+
+    # Teradata uses the TOP statement instead of LIMIT which is what SQL92 uses
+    # TOP is expected after DISTINCT and not at the end of the query
+    # e.g: SELECT TOP 100 * FROM my_table
+    if (!is.null(limit) && !identical(limit, Inf)) {
+      assert_that(is.numeric(limit), length(limit) == 1L, limit > 0)
+      build_sql(" TOP ", as.integer(limit), " ")},
+
+    escape(select, collapse = ", ", con = con)
+  )
+
+  out$from      <- sql_clause_from(from, con)
+  out$where     <- sql_clause_where(where, con)
+  out$group_by  <- sql_clause_group_by(group_by, con)
+  out$having    <- sql_clause_having(having, con)
+  out$order_by  <- sql_clause_order_by(order_by, con)
+
+
+  escape(unname(compact(out)), collapse = "\n", parens = FALSE, con = con)
+}
+
+#' @export
+sql_translate_env.Teradata <- function(con) {
+  sql_variant(
+    sql_translator(.parent = base_odbc_scalar,
+      as.numeric    = sql_cast("NUMERIC"),
+      as.double     = sql_cast("NUMERIC"),
+      as.character  = sql_cast("VARCHAR(MAX)"),
+      var           = sql_prefix("VAR_SAMP"),
+      log10         = sql_prefix("LOG"),
+      log           = function(x, base = exp(1)) {
+                        if (isTRUE(all.equal(base, exp(1)))) {
+                          build_sql("ln(", x, ")")
+                        } else {
+                          # Use log change-of-base because postgres doesn't support the
```

Comment:
Is this true for teradata too? If so we should pull out into a resuable function

Reply:
Ok, where should I place the function? I'm thinking `traslate-sql-helpers.R`

Reply:
Yeah, sounds good.